spock-core/src/main/java/org/spockframework/util/ReflectionUtil.java

```diff
@@ -38,30 +53,88 @@ public static String getPackageName(Class<?> clazz) {
   public static Class<?> loadFirstAvailableClass(String... classNames) {
     for (String className : classNames) {
       Class<?> clazz = loadClassIfAvailable(className);
-      if (clazz != null) {return clazz;}
+      if (clazz != null) {
+        return clazz;
+      }
     }
     return null;
   }
 
   public static Class<?> loadClassIfAvailable(String className) {
     try {
-      return ReflectionUtil.class.getClassLoader().loadClass(className);
+      return loadClass(className);
     } catch (ClassNotFoundException e) {
       return null;
     }
   }
 
+  /**
+   * Loads the className in the following order:
+   * <ul>
+   * <li>Spock {@link ClassLoader}</li>
+   * <li>{@code Thread.currentThread().getContextClassLoader()}</li>
+   * </ul>
+   *
+   * @param className the class to load
+   * @return the loaded class
+   * @throws ClassNotFoundException if the class could not be found
+   */
+  public static Class<?> loadClass(String className) throws ClassNotFoundException {
```

Comment:
Can't this lead to classloader leaks and other visibility problems, if the life time of the caller/user exceeds that of the classloader?

Reply:
I don't understand exactly what you mean. Can you elaborate your concerns?

The lifetime of a classloader depends on the classes it has directly loaded , not transitivly though another classloader, e.g. the `Class.getClassLoader()`.
So the caller of this methods can only "leak" the classloader, if it would also not unref the returned Class instance, which would also happen in the old code.